### PR TITLE
RSDK-9338: add optional parameters in data client

### PR DESCRIPTION
**Pull request by purplenicole730**
Changes:
- rename some structs for clarity
- add optional parameter structs
- add comments clarifying what happens if a parameter is left empty

Note as to why I didn't put all optional parameters in structs: 
At first, I thought I should be putting the optional `filter` parameters for some of the methods in its own `<method>Options` struct. Thinking over it some more, I realized these functions could be seen as ones where the user is purposely leaving this parameter `nil`. This is in line with my discussions with Benji when we first introduced optional parameter structs. It also seemed more straightforward, especially with more clarifying comments.

### CODE REVIEW

File: app/data_client.go [hunk 1]
```diff
@@ -235,26 +235,41 @@ type FileData struct {
...
 }
 
-// BinaryOptions represents optional parameters for the BinaryDataCaptureUpload method.
-type BinaryOptions struct {
+// DataByFilterOptions contains optional parameters for TabularDataByFilter and BinaryDataByFilter.
+type DataByFilterOptions struct {
+	// No Filter implies all data.
+	Filter *Filter
+	// Limit defaults to 50 if unspecified.
+	Limit int
```
Comment (jckras) [inline app/data_client.go:243]: if Limit is not a pointer, how do we know whether they set the limit to 0 or whether they did not set it?

File: app/data_client.go [hunk 2]
```diff
@@ -320,22 +346,23 @@ func BsonToGo(rawData [][]byte) ([]map[string]interface{}, error) {
...
-	sortOrder Order,
-	countOnly bool,
-	includeInternalData bool,
-) (TabularDataReturn, error) {
+func (d *DataClient) TabularDataByFilter(ctx context.Context, opts *DataByFilterOptions) (TabularDataReturn, error) {
+	dataReq := pb.DataRequest{}
+	var countOnly, includeInternalData bool
+	if opts != nil {
+		dataReq.Filter = filterToProto(opts.Filter)
+		if opts.Limit != 0 {
```
Comment (jckras) [inline app/data_client.go:354]: If `opts.Limit` is not specified, should we default it to 50 somewhere in this function? also, how do we differentiate between a limit explicitly set to 0 and one that is not set (given that the default value for integers in Go is 0, iirc)?

**Review by jckras — APPROVED** [review]
left a question about the Limit parameter for `DataByFilter` functions otherwise LGTM! thanks for doing this nicole!

File: app/data_client.go (hunk 1 above)
Comment (purplenicole730) [reply app/data_client.go:243]: A user would not put the limit to 0, or they would not get any data at all.

File: app/data_client.go (hunk 2 above)
Comment (purplenicole730) [reply app/data_client.go:354]: This is something I know from my work in the Python SDK. The default to 50 happens server-side

File: app/data_client.go (hunk 1 above)
Comment (purplenicole730) [reply app/data_client.go:243]: The limit is there to limit how many entries to include in a page of data.

**Pull request merged**